firedrake/dmplex.pyx

```diff
@@ -819,7 +835,16 @@ def cell_facet_labeling(PETSc.DM plex,
             if fstart <= point < fend:
                 # This is a facet point
                 DMLabelHasPoint(label, point, &is_interior)
-                cell_facets[cell, fi] = <np.int8_t>is_interior
+                cell_facets[cell, fi, 0] = <np.int8_t>is_interior
+
+                # Collect subdomains (if any)
+                if exterior_subdomains:
+                    # If on an exterior facet, search for the subdomain marker
+                    if not is_interior:
+                        for marker in exterior_markers:
+                            # Facet point in subdomain
+                            if point in sub_domain_ids[marker].array:
+                                cell_facets[cell, fi, 1] = <np.int8_t>marker
```

Comment:
This is terrible.

1. you've put python calls in the inner loop
2. For every point, you search the whole array, turning this into an (at best) NlogN, but probably N^2 code path.

Just write:
```python
cell_facets[cell, fi, 0] = <np.int8_t>is_interior
if not is_interior:
    CHKERR(DMLabelGetValue(label, point, &marker)
    cell_facets[cell, fi, 1] = <np.int8_t>marker
```

Reply:
Also, it is legitimate for some exterior facets to be unlabelled (DMLabelGetValue returns -1).  Similarly, I see for now you don't handle interior facet subdomains (although that is also allowed).

Reply:
>This is terrible.

Yeah I don't like it either. That's why we have peer-reviewed code though, right? Someone might know something I'm missing here.

>Just write:
>```python
>cell_facets[cell, fi, 0] = <np.int8_t>is_interior
>if not is_interior:
 >   CHKERR(DMLabelGetValue(label, point, &marker)
 >   cell_facets[cell, fi, 1] = <np.int8_t>marker
>```

Great, this is helpful. I'm assuming the `label` here is NOT the same one used to determine facet types (exterior/interior). So something like:
```python
cdef:
    ...
    PetscInt marker
    DMLabel sub_domain
    ...
CHKERR(DMGetLabel(plex.dm, FACE_SETS_LABEL.encode(), &sub_domain))
...
CHKERR(DMLabelGetValue(sub_domain, point, &marker))
cell_facets[cell, fi, 1] = <np.int8_t>marker
```
Or is there a way I can reuse the label to do this? Note that if the mesh is spherical, then `CHKERR(DMGetLabel(plex.dm, FACE_SETS_LABEL.encode(), &sub_domain))` together with `CHKERR(DMLabelGetValue(sub_domain, point, &marker))` causes a segmentation violation. This error can also be reproduced if you simply run:
```python
from firedrake import *
mesh = UnitIcosahedralSphereMesh(2)
mesh.init_cell_orientations(Expression(("x[0]", "x[1]", "x[2]")))
plex = mesh._plex
plex.getLabelIdIS("Face Sets").allGather()
```
Is this a bug/expected behaviour or did I do something wrong? If I don't special case sphere meshes, then at some point this causes tests on spheres to break due to a `segv` error. Perhaps by checking:
```python
if plex.getLabelIdIS("exterior_facets").size == 0:
    facet_set = "interior_facets"
else:
    facet_set = FACE_SETS_LABEL
CHKERR(DMGetLabel(plex.dm, facet_set.encode(), &sub_domain))
...
```
Basically, I guess I want to check if the mesh has a boundary. This was just one way I thought I could do this. Perhaps I don't need to? I'll push an update anyway and we'll go from there.
>Also, it is legitimate for some exterior facets to be unlabelled (DMLabelGetValue returns -1). Similarly, I see for now you don't handle interior facet subdomains (although that is also allowed).

OK. I see no reason why I shouldn't just handle interior facet subdomains then. So I'll make some modifications.